addon/utils.js

```diff
@@ -440,20 +440,26 @@ export function removeInternalValues (val) {
 
   if (!Array.isArray(val)) {
     let withoutInternal
+    let merge
+    let without
     if (immutable.isImmutable(val)) {
-      withoutInternal = val.without('_internal')
-      const childrenWithoutInternals = removeChildInternals(withoutInternal)
-      return withoutInternal.merge(childrenWithoutInternals)
-    }
-
-    if (val._internal !== undefined) {
-      withoutInternal = Object.assign({}, val)
-      delete withoutInternal._internal
+      merge = immutable.merge
+      without = immutable.without
+    } else if (val._internal !== undefined) {
+      merge = Object.assign
+      without = _.omit
     } else {
-      withoutInternal = val
+      merge = function (first, second) {
```

Comment:
Would this local `merge` implementation be better suited in external to this function. Nested functions do have some performance drawbacks especially inside something that will be recursed. A new lexical environment (closure) needs to be created for every recursive step.